### ISSUE TITLE
/user/uploads: Sporadically returns []deals with null dealId and dates

### ISSUE THREAD

**Issue by 3456091** [web3-storage/web3.storage#1512]
I'm using the [Go client](https://github.com/web3-storage/go-w3s-client) and sporadically seeing errors returned by the Client.List() method. The error appears to emanate from decoding the JSON received when a time.Time value is expected but an empty string is received. Unfortunately, the errors aren't wrapped and this behaviour is sporadic so I'm yet to reproduce.

Looking at [status.go](https://github.com/web3-storage/go-w3s-client/blob/v0.0.6/status.go), some dates are parsed like this:

```
if raw.Activation != "" {
    d.Activation, err = time.Parse(iso8601, raw.Activation)
    if err != nil {
        return err
   }
}
```

i.e. they allow for the date field to be missing. The "deal" JSON assumes the _created_ and _updated_ fields will never be empty, and the "status" JSON assumes the _created_ field will never be empty. So, I'm assuming one or more of these fields are implicated in this bug.

Any thoughts on whether the problem is with the API or the client? The Go client can easily be patched to ignore these missing dates, but should they be missing in the first place?

**Comment by 3456091** [comment]
(Last time I saw a date error like this in the Go client was https://github.com/web3-storage/go-w3s-client/issues/21 , and the problem was in the "deals" array).

**Comment by 3456091** [comment]
I've managed to recreate. The /user/uploads endpoint returns an array member like the following:

```
{
  "_id": "215318962269329065",
  "type": "Car",
  "name": "Upload at 2022-06-19T13:12:27.705Z",
  "created": "2022-06-19T13:12:27.705+00:00",
  "updated": "2022-06-19T13:12:27.705+00:00",
  "cid": "bafybeibv6qpogx456s6ub7ems2yhtcevldw367hvjqiqxkdg4nr76m6tly",
  "dagSize": 1049766,
  "pins": [
    {
      "status": "Pinned",
      "updated": "2022-06-19T13:34:22.553945+00:00",
      "peerId": "12D3KooWSafoW6yrSL7waghFAaiCqGy5mdjpQx4jn4CRNqbG7eqG",
      "peerName": "web3-storage-am6-8",
      "region": null
    },
    {
      "status": "Pinned",
      "updated": "2022-06-19T13:34:22.553945+00:00",
      "peerId": "12D3KooWDRak1XzURGh9MvGR4EWaP9kcbmdoagAcGMcNxBXXLzTF",
      "peerName": "web3-storage-dc13-8",
      "region": null
    },
    {
      "status": "Pinned",
      "updated": "2022-06-19T13:34:22.553945+00:00",
      "peerId": "12D3KooWSH5uLrYe7XSFpmnQj1NCsoiGeKSRCV7T5xijpX2Po2aT",
      "peerName": "web3-storage-sv15-8",
      "region": null
    }
  ],
  "deals": [
    {
      "dealId": null,
      "storageProvider": null,
      "status": "Queued",
      "pieceCid": "baga6ea4seaql2leesalhfwmb5xbvhrotu4x76mxn25wcpile6qpbn777omdmqfa",
      "dataCid": "bafybeieesr6dknelbduxqmmtt7q2nernm4mmka4k66t7f4d6sgsq3ixc4u",
      "dataModelSelector": "Links/144/Hash/Links/2/Hash/Links/0/Hash",
      "activation": null,
      "expiration": null,
      "created": null,
      "updated": null
    }
  ]
}
```

Note that a deal exists, but most of its key fields are null: this is what's crashing the Go client. I suspect it happens when /user/uploads is requested very soon after an upload. I'm assuming this should be fixed on the backend?

**Comment by vasco-santos** [comment]
I think there are two different things here.

First, we have a fault tolerance logic in the back end to not fail if we get a timeout getting the deals: https://github.com/web3-storage/web3.storage/blob/main/packages/db/index.js#L928

main reason is because deals are in a Foreign Database and queries sometimes take longer than our postgrest heroku dyno timeout (we need to look into that definitely!).

However, go client should be able to handle either empty deals, or deals with null fields. Would you like to PR a fix for this part?

**Comment by 3456091** [comment]
> First, we have a fault tolerance logic in the back end to not fail if we get a timeout getting the deals: https://github.com/web3-storage/web3.storage/blob/main/packages/db/index.js#L928
> 
> main reason is because deals are in a Foreign Database and queries sometimes take longer than our postgrest heroku dyno timeout (we need to look into that definitely!).

OK, perhaps you could open a new issue for that, then we close this one?

> However, go client should be able to handle either empty deals, or deals with null fields. Would you like to PR a fix for this part?

Done: https://github.com/web3-storage/go-w3s-client/pull/27

**Comment by alanshaw** [comment]
Thank you - released in `v0.0.7` of the `go-w3s-client` library.
